ROOT TGeoMatrix interface change

Dear DD4Hep developers,

dd4hep version 01-05 (and, I think, master as well, since the code in question is unchanged) fails to build with gcc 6.2 (or newer) and HEAD version of ROOT. The error message is "invalid initialization of non-const reference of type ‘TGeoMatrix&’ from an rvalue of type ‘TGeoMatrix’". 

Thanks,
Ivan for the GENSER team

 - OS version: SLC6 (not tested, but will likely fail with Centos 7)
 - Compiler version: GCC 6.2, 7.3.0, 8.1.0
 - DD4hep version: tag v01-05
 - Reproduced by: building against ROOT-head.
 - Build log: https://cernbox.cern.ch/index.php/s/TxSdekx7XSLl9bA
Important part:
```
[ 44%] Building CXX object UtilityApps/CMakeFiles/geoConverter.dir/src/converter.cpp.o
/mnt/build/jenkins/workspace/lcg_personal_experimental/BUILDTYPE/Release/COMPILER/gcc62binutils/LABEL/slc6/build/projects/DD4hep-01-05/src/DD4hep/01-05/DDCore/src/plugins/LCDDConverter.cpp: In member function ‘virtual xml_h dd4hep::detail::LCDDConverter::handleSolid(const string&, const TGeoShape*) const’:
/mnt/build/jenkins/workspace/lcg_personal_experimental/BUILDTYPE/Release/COMPILER/gcc62binutils/LABEL/slc6/build/projects/DD4hep-01-05/src/DD4hep/01-05/DDCore/src/plugins/LCDDConverter.cpp:552:39: error: invalid initialization of non-const reference of type ‘TGeoMatrix&’ from an rvalue of type ‘TGeoMatrix’
         TGeoMatrix& linv = lm->Inverse()
                            ~~~~~~~~~~~^~
/mnt/build/jenkins/workspace/lcg_personal_experimental/BUILDTYPE/Release/COMPILER/gcc62binutils/LABEL/slc6/build/projects/DD4hep-01-05/src/DD4hep/01-05/DDCore/src/plugins/LCDDConverter.cpp:571:39: error: invalid initialization of non-const reference of type ‘TGeoMatrix&’ from an rvalue of type ‘TGeoMatrix’
         TGeoMatrix& rinv = rm->Inverse()
                            ~~~~~~~~~~~^~
[ 44%] Building CXX object DDG4/CMakeFiles/DDG4.dir/src/Geant4Exec.cpp.o
[ 44%] Building CXX object DDG4/CMakeFiles/DDG4.dir/src/Geant4Field.cpp.o
[ 44%] Building CXX object DDCond/CMakeFiles/DDCond.dir/src/ConditionsOperators.cpp.o
[ 44%] Building CXX object DDEve/CMakeFiles/DDEve.dir/src/DDEvePlugins.cpp.o
make[6]: *** [DDCore/CMakeFiles/DDCorePlugins.dir/src/plugins/LCDDConverter.cpp.o] Error 1
```

When I build with gcc7 still has this 
```
/build/mrodozov/builds/build/update_root/build_dd4hep_with_latest_root614/BUILD/slc6_amd64_gcc700/external/dd4hep/v01-08x-cms/dd4hep-v01-08x/DDCore/src/plugins/Compact2Objects.cpp:1118:101: error: taking ad\
dress of temporary [-fpermissive]
   Transform3D  tr_volume(detail::matrix::_transform(&anchor.nominal().worldTransformation().Inverse()));
```

--- a/DDCore/include/DD4hep/MatrixHelpers.h
+++ b/DDCore/include/DD4hep/MatrixHelpers.h
@@ -63,6 +63,7 @@ namespace dd4hep {
 
       /// Convert a TGeoMatrix object to a generic Transform3D                                  \ingroup DD4HEP \ingroup DD4HEP_CORE
       Transform3D      _transform(const TGeoMatrix* matrix);
+      Transform3D      _transform(const TGeoMatrix& matrix);
 
       /// Decompose a generic Transform3D into a translation (Position) and a RotationZYX       \ingroup DD4HEP \ingroup DD4HEP_CORE
       void _decompose(const Transform3D& trafo, Position& pos, RotationZYX& rot);

--- a/DDCore/src/MatrixHelpers.cpp
+++ b/DDCore/src/MatrixHelpers.cpp
@@ -112,6 +112,19 @@ Transform3D dd4hep::detail::matrix::_transform(const TGeoMatrix* matrix)    {
                      0e0,0e0,0e0,t[1]*MM_2_CM,
                      0e0,0e0,0e0,t[2]*MM_2_CM);
 }
+// add another implementation that takes const reference
+Transform3D dd4hep::detail::matrix::_transform(const TGeoMatrix& matrix)    {
+  const Double_t* t = matrix.GetTranslation();
+  if ( matrix.IsRotation() )  {
+    const Double_t* r = matrix.GetRotationMatrix();
+    return Transform3D(r[0],r[1],r[2],t[0]*MM_2_CM,
+                       r[3],r[4],r[5],t[1]*MM_2_CM,
+                       r[6],r[7],r[8],t[2]*MM_2_CM);
+  }
+  return Transform3D(0e0,0e0,0e0,t[0]*MM_2_CM,
+                     0e0,0e0,0e0,t[1]*MM_2_CM,
+                     0e0,0e0,0e0,t[2]*MM_2_CM);
+}
 
 dd4hep::XYZAngles dd4hep::detail::matrix::_xyzAngles(const TGeoMatrix* m) {
   return m->IsRotation() ? _xyzAngles(m->GetRotationMatrix()) : XYZAngles(0,0,0);

--- a/DDCore/src/plugins/Compact2Objects.cpp
+++ b/DDCore/src/plugins/Compact2Objects.cpp
@@ -1115,7 +1115,7 @@ template <> void Converter<Parallelworld_Volume>::operator()(xml_h element) cons
   }
 
   /// Create the shape and the corresponding volume
-  Transform3D  tr_volume(detail::matrix::_transform(&anchor.nominal().worldTransformation().Inverse()));
+  Transform3D  tr_volume(detail::matrix::_transform(anchor.nominal().worldTransformation().Inverse()));
   Solid        sol(shape.createShape());
   Volume       vol(name, sol, mat);
   Volume       par = conn ? description.worldVolume() : description.parallelWorldVolume();

--- a/DDCore/src/plugins/LCDDConverter.cpp
+++ b/DDCore/src/plugins/LCDDConverter.cpp
@@ -549,7 +549,7 @@ xml_h LCDDConverter::handleSolid(const string& name, const TGeoShape* shape) con
         obj.setAttr(_U(unit), "cm");
       }
       if (lm->IsRotation()) {
-        TGeoMatrix& linv = lm->Inverse();
+        TGeoMatrix const & linv = lm->Inverse();
         XYZRotation  rot = getXYZangles(linv.GetRotationMatrix());
         if ((rot.X() != 0.0) || (rot.Y() != 0.0) || (rot.Z() != 0.0)) {
           first_solid.append(obj = xml_elt_t(geo.doc, _U(firstrotation)));
@@ -568,7 +568,7 @@ xml_h LCDDConverter::handleSolid(const string& name, const TGeoShape* shape) con
         solid.setRef(_U(positionref), pos.name());
       }
       if (rm->IsRotation()) {
-        TGeoMatrix& rinv = rm->Inverse();
+        TGeoMatrix const & rinv = rm->Inverse();
         XYZRotation rot = getXYZangles(rinv.GetRotationMatrix());
         if ((rot.X() != 0.0) || (rot.Y() != 0.0) || (rot.Z() != 0.0)) {
           xml_ref_t xml_rot = handleRotation(rnam+"_rot", &rinv);

